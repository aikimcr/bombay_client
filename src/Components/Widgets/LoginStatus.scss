@import '../../global.scss';

.login-status {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: var(--header-bar-parts-background-color);
    width: 145px;
    margin-left: 5px;
    flex-grow: 0;
}