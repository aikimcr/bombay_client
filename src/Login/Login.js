import { useEffect, useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import './Login.scss';

import BombayLoginContext from '../Context/BombayLoginContext';
import BombayModeContext from '../Context/BombayModeContext';
import BombayUtilityContext from '../Context/BombayUtilityContext';

import { login } from "../Network/Login";
import LabeledInput from '../Widgets/LabeledInput';

function Login(props) {
    const navigate = useNavigate();

    let timerHandle = null;

    const [error, setError] = useState(null);

    const loggedIn = useContext(BombayLoginContext);
    const mode = useContext(BombayModeContext);
    const { checkLoginState } = useContext(BombayUtilityContext);

    useEffect(() => {
        if (loggedIn) {
            navigate(`/${mode}`);
        }
    }, [loggedIn, mode, navigate]);

    function getCredentials() {
        // TODO: THere has to be a better way to do what these two lines are doing.
        const usernameInput = document.querySelector('[data-fieldname="username"] input');
        const passwordInput = document.querySelector('[data-fieldname="password"] input');

        return { username: usernameInput.value, password: passwordInput.value };
    }

    function clearTimer() {
        if (timerHandle) {
            clearTimeout(timerHandle);
            timerHandle = null;
        }
    }

    function handleChange(event) {
        timerHandle = setTimeout(() => {
            const loginButton = document.querySelector('.btn.login');

            clearTimer();

            const { username, password } = getCredentials();

            if (username.length > 0 && password.length > 0) {
                loginButton.classList.remove('disabled');
            } else {
                loginButton.classList.add('disabled');
            }
        }, 250);
    }

    function clearAllFields(event) {
        const loginButton = document.querySelector('.btn.login');

        clearTimer();

        document.querySelector('[data-fieldname="username"] input').value = '';
        document.querySelector('[data-fieldname="password"] input').value = '';
        loginButton.classList.add('disabled');
        setError(null);
    }

    async function doLogin(event) {
        const { username, password } = getCredentials();

        clearTimer();

        await login(username, password)
            .catch(err => {
                if (err.status === 401) {
                    setError('Username or password is incorrect.');
                } else {
                    setError(`${err.status}: ${err.message}`);
                }
            });

        checkLoginState();
    }

    if (loggedIn) {
        return '';
    } else {
        return (
            <div className="login-container">
                <div className="login-form">
                    <h1 className="login-header">Please Log In</h1>
                    <div className="info">
                        <LabeledInput
                            modelName='login'
                            fieldName='username'
                            labelText='User Name'
                            onChange={handleChange}
                        />
                        <LabeledInput
                            modelName='login'
                            fieldName='password'
                            labelText='Password'
                            type='password'
                            onChange={handleChange}
                        />
                    </div>
                    {error && <div className="error">{error}</div>}
                    <div className="controls">
                        <div className="clear btn" onClick={clearAllFields}>Clear All Fields</div>
                        <div className="login btn disabled" onClick={doLogin}>Login</div>
                    </div>
                </div>
            </div>
        );
    }
}

export default Login;