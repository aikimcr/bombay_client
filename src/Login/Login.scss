@import '../global.scss';

.login-header {
    text-align: center;
}

.login-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: black;
}

.login-form {
    width: 600px;
    height: 400px;
    position: relative;
    top: calc(50% - 200px);
    left: calc(50% - 300px);
    background: white;
    padding-top: 60px;

    .info>div {
        width: 400px;
        display: grid;
        grid-template-columns: 100px 300px;
    }

    .info,
    .error,
    .controls {
        position: relative;
        left: 100px;
        width: 400px;
        margin-top: 20px;
    }

    .error {
        border: thin solid red;
        color: red;
        font-weight: bold;
        padding: 10px;
    }

    .controls {
        display: flex;
        justify-content: space-between;
    }

}

// Crossed out eye "\f070"; "\01F441";
// Eye "\f06e";