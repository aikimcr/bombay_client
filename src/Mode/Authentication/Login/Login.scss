@import '../../../global.scss';

.login-header {
    text-align: center;
}

.login-container {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
    width: 100%;
    height: 100%;
    background: transparent;
}

.login-form {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: start;
    gap: 1rem;
    width: 100%;
    height: 100%;
    background: transparent;
    @include form-container;

    .info {
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        align-items: flex-start;
        gap: 1rem;
        width: 100%;
        height: 100%;
    }

    .error {
        border: thin solid red;
        color: red;
        font-weight: bold;
        padding: 10px;
    }

    .controls {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: flex-start;
        flex-wrap: nowrap;
        gap: 1rem;
        width: 100%;
        height: 100%;
    }

}

// Crossed out eye "\f070"; "\01F441";
// Eye "\f06e";