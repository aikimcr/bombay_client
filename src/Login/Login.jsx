import { useContext, useState } from "react";
import { createPortal } from "react-dom";

import "./Login.scss";

import BombayLoginContext from "../Context/BombayLoginContext";

import { loginStatus, login } from "../Network/Login";
import OldLabeledInput from "../Components/Widgets/OldLabeledInput";
import Button from "../Components/Widgets/Button";

function Login(props) {
  let timerHandle = null;

  const [error, setError] = useState(null);
  const [submitDisabled, setSubmitDisabled] = useState(true);

  const { loggedIn, setLoggedIn, showLoginForm } =
    useContext(BombayLoginContext);

  function getCredentials() {
    // TODO: THere has to be a better way to do what these two lines are doing.
    const usernameInput = document.querySelector(
      '[data-targetfield="username"] input',
    );
    const passwordInput = document.querySelector(
      '[data-targetfield="password"] input',
    );

    return { username: usernameInput.value, password: passwordInput.value };
  }

  function clearTimer() {
    if (timerHandle) {
      clearTimeout(timerHandle);
      timerHandle = null;
    }
  }

  function handleChange(event) {
    timerHandle = setTimeout(() => {
      clearTimer();

      const { username, password } = getCredentials();

      if (username.length > 0 && password.length > 0) {
        setSubmitDisabled(false);
      } else {
        setSubmitDisabled(true);
      }
    }, 250);
  }

  function clearAllFields(event) {
    clearTimer();

    document.querySelector('[data-targetfield="username"] input').value = "";
    document.querySelector('[data-targetfield="password"] input').value = "";
    setSubmitDisabled(true);
    setError(null);
  }

  async function doLogin(event) {
    const { username, password } = getCredentials();

    clearTimer();

    await login(username, password).catch((err) => {
      if (err.status === 401) {
        setError("Username or password is incorrect.");
      } else {
        setError(`${err.status}: ${err.message}`);
      }
    });

    const loginOkay = await loginStatus();
    setLoggedIn(loginOkay);
  }

  if (loggedIn) return null;
  if (!showLoginForm) return null;

  const modalRoot = document.getElementById("modal-root");

  return createPortal(
    <div className="login-container">
      <div className="login-form">
        <h1 className="login-header">Please Log In</h1>
        <div className="info">
          <OldLabeledInput
            modelName="login"
            fieldName="username"
            labelText="User Name"
            onChange={handleChange}
          />
          <OldLabeledInput
            modelName="login"
            fieldName="password"
            labelText="Password"
            type="password"
            onChange={handleChange}
          />
        </div>
        {error && <div className="error">{error}</div>}
        <div className="controls">
          <Button
            className="clear"
            disabled={false}
            onClick={clearAllFields}
            label="Clear All Fields"
          />
          <Button
            className="login"
            disabled={submitDisabled}
            onClick={doLogin}
            label="Login"
          />
        </div>
      </div>
    </div>,
    modalRoot,
  );
}

export default Login;
