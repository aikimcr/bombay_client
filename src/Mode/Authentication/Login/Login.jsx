import React, { useContext, useEffect, useRef, useState } from "react";
import { useForm } from "react-hook-form";
import BombayLoginContext from "../../../Context/BombayLoginContext";
import { loginStatus, login } from "../../../Network/Login";
import {
  Button,
  LabeledInput,
  PasswordInput,
  TextInput,
} from "../../../Components";
import { useRouteManager } from "../../../Hooks/useRouteManager";

import "./Login.scss";

export const Login = () => {
  const routeManager = useRouteManager();

  const { loggedIn, setLoggedIn } = useContext(BombayLoginContext);

  const [submitDisabled, setSubmitDisabled] = useState(true);

  const { register, reset, handleSubmit, formState, setError, clearErrors } =
    useForm({
      shouldUseNativeValidaton: true,
    });

  const doLogin = async (data) => {
    const { username, password } = data;

    try {
      await login(username, password);
    } catch (err) {
      if (err.status === 401) {
        setError("Username or password is incorrect.");
      } else {
        setError(`${err.status}: ${err.message}`);
      }
    }

    const loginOkay = await loginStatus();
    setLoggedIn(loginOkay);
  };

  const onSubmit = (data) => {
    doLogin(data);
  };

  const handleReset = () => {
    setSubmitDisabled(true);
    reset({
      username: "",
      password: "",
    });
  };

  useEffect(() => {
    if (formState.isValid && Object.entries(formState.errors).length === 0) {
      setSubmitDisabled(false);
      return;
    }

    setSubmitDisabled(true);
  }, [formState]);

  useEffect(() => {
    if (loggedIn) {
      routeManager.navigateToRoute("/artistList");
    }
  }, [loggedIn]);

  return (
    <div className="login-container">
      <form
        className="login-form"
        onSubmit={handleSubmit(onSubmit)}
        onReset={handleReset}
      >
        <h1 className="login-header">Please Log In</h1>
        <div className="info">
          <LabeledInput
            labelText="User Name"
            inputId="login-user-name"
            InputField={TextInput}
            inputProps={{
              ...register("username", {
                required: "Please enter a valid username",
              }),
              id: "login-user-name",
              name: "username",
            }}
          />
          <LabeledInput
            labelText="Password"
            inputId="login-password"
            InputField={PasswordInput}
            inputProps={{
              ...register("password", {
                required: "Please enter a valid password",
              }),
              id: "login-password",
              name: "password",
            }}
          />
        </div>
        {Object.entries(formState.errors).length !== 0 && (
          <div className="error">{formState.errors[0]}</div>
        )}
        <div className="controls">
          <Button
            className="clear"
            disabled={false}
            role="secondary"
            text="Clear All Fields"
            type="reset"
          />
          <Button
            className="login"
            disabled={submitDisabled}
            text="Login"
            type="submit"
          />
        </div>
      </form>
    </div>
  );
};
