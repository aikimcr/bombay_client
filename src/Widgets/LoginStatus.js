import { useContext } from 'react';

import './LoginStatus.scss';

import BombayLoginContext from '../Context/BombayLoginContext';
import BombayUtilityContext from '../Context/BombayUtilityContext';

import { logout } from "../Network/Login";
import Button from './Button';

function LoginStatus(props) {
    const loggedIn = useContext(BombayLoginContext);
    const { setLoginState } = useContext(BombayUtilityContext);

    async function doLogout() {
        await logout();
        setLoginState(false);
    }

    return (
        <div className="login-status">
            {loggedIn ? <Button className="logout btn" onClick={doLogout} disabled={false} label='Logout' /> : <div className="label">Please Login</div>}
        </div>
    );
}

export default LoginStatus;