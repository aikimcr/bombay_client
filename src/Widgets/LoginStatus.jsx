import { useContext } from "react";

import "./LoginStatus.scss";

import BombayLoginContext from "../Context/BombayLoginContext";

import { logout } from "../Network/Login";
import Button from "./Button.jsx";

function LoginStatus(props) {
  const { loggedIn, setLoggedIn, setShowLogin } =
    useContext(BombayLoginContext);

  async function doLogout() {
    await logout();
    setLoggedIn(false);
  }

  function showLogin() {
    setShowLogin(true);
  }

  return (
    <div className="login-status">
      {loggedIn ? (
        <Button
          className="logout btn"
          onClick={doLogout}
          disabled={false}
          label="Logout"
        />
      ) : (
        <Button
          className="login btn"
          onClick={showLogin}
          disabled={false}
          label="Login"
        />
      )}
    </div>
  );
}

export default LoginStatus;
