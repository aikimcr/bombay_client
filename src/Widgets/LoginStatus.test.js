import * as NetworkLogin from "../Network/Login";
jest.mock("../Network/Login");

import { act, render, screen } from "@testing-library/react";

import LoginStatus from "./LoginStatus";

beforeEach(() => {
  localStorage.removeItem("jwttoken");
});

afterEach(() => {
  localStorage.removeItem("jwttoken");
});

it("should show the login button when logged out", async () => {
  const { asFragment } = render(
    <ContextChanger loggedIn={false}>
      <LoginStatus />
    </ContextChanger>,
  );

  expect(asFragment).toMatchSnapshot();

  const loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).toBeNull();

  const loginButton = screen.getByText("Login");
  expect(loginButton).toBeVisible();
  expect(loginButton.tagName).toBe("BUTTON");
});

it("should show the login form when the login button is pressed", async () => {
  render(
    <ContextChanger loggedIn={false}>
      <LoginStatus />
    </ContextChanger>,
  );

  let loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).toBeNull();

  const loginButton = screen.getByText("Login");

  await act(async () => {
    loginButton.click();
  });

  loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).not.toBeNull();
  expect(loginForm).toBeInTheDocument();
});

it("should show the logout button when logged in", async () => {
  const { asFragment } = render(
    <ContextChanger loggedIn={true}>
      <LoginStatus />
    </ContextChanger>,
  );
  expect(asFragment).toMatchSnapshot();

  const loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).toBeNull();

  const logoutButton = screen.getByText("Logout");
  expect(logoutButton).toBeVisible();
  expect(logoutButton.tagName).toBe("BUTTON");
});

it("should logout when the logout button is pressed", async () => {
  const { resolve } = NetworkLogin._setupMocks();

  const result = render(
    <ContextChanger loggedIn={true}>
      <LoginStatus />
    </ContextChanger>,
  );

  let loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).toBeNull();

  const logoutButton = screen.getByText("Logout");

  await act(async () => {
    logoutButton.click();
    resolve({});
  });

  expect(NetworkLogin.logout).toBeCalledTimes(1);

  loginForm = screen.queryByText("Showing Log In Form");
  expect(loginForm).toBeNull();
});
