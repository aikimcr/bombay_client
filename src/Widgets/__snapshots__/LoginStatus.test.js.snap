// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the login status (logged out) 1`] = `[Function]`;

exports[`should show the logout button when logged in 1`] = `[Function]`;
